### PR TITLE
reenable compiled method cache

### PR DESCRIPTION
**Pull request by mykmelez**
This branch reenables the compiled method cache, using hashes for ClassInfo and MethodInfo IDs to maintain their stability across runs. The cache reduces startup time significantly relative to compilation without caching:

|         Test         | Baseline Mean |     Mean    |     +/-     |     %    |     P    |      Min     |      Max      | 
|:---------------------|--------------:|------------:|------------:|---------:|---------:|-------------:|--------------:|
| startupTime          |       1,265ms |       976ms |      -289ms |   -22.86 |   BETTER |        964ms |       1,008ms | 
| vmStartupTime        |         212ms |       206ms |        -6ms |    -3.02 |     SAME |        186ms |         243ms | 
| bgStartupTime        |          52ms |        43ms |        -9ms |   -17.10 |   BETTER |         41ms |          44ms | 
| fgStartupTime        |         848ms |       612ms |      -236ms |   -27.81 |   BETTER |        591ms |         666ms | 
| fgRefreshStartupTime |         153ms |       115ms |       -38ms |   -24.93 |   BETTER |         88ms |         121ms | 
| fgRestartTime        |           n/a |       NaNms |         n/a |      n/a |      n/a |   Infinityms |   -Infinityms | 
| *totalSize*          |    *61,654kb* |  *59,026kb* |  *-2,628kb* |  *-4.26* | *BETTER* |   *57,800kb* |    *59,859kb* | 
| *domSize*            |       *169kb* |     *169kb* |       *0kb* |   *0.00* |   *SAME* |      *168kb* |       *169kb* | 
| *styleSize*          |       *397kb* |     *396kb* |       *0kb* |  *-0.08* |   *SAME* |      *395kb* |       *398kb* | 
| *jsObjectsSize*      |    *51,445kb* |  *51,387kb* |     *-58kb* |  *-0.11* | *BETTER* |   *51,368kb* |    *51,391kb* | 
| *jsStringsSize*      |       *541kb* |     *499kb* |     *-42kb* |  *-7.75* | *BETTER* |      *497kb* |       *501kb* | 
| *jsOtherSize*        |     *8,866kb* |   *6,338kb* |  *-2,528kb* | *-28.51* | *BETTER* |    *5,115kb* |     *7,166kb* | 
| *otherSize*          |       *237kb* |     *236kb* |       *0kb* |  *-0.20* |   *SAME* |      *224kb* |       *239kb* | 
| *USS*                |         *n/a* |     *NaNkb* |       *n/a* |    *n/a* |    *n/a* | *Infinitykb* | *-Infinitykb* | 
| *peakRSS*            |   *509,009kb* | *458,217kb* | *-50,792kb* |  *-9.98* | *BETTER* |  *453,832kb* |   *458,748kb* | 

In fact startup time on this branch with compilation enabled is now better than startup time on master:

|         Test         | Baseline Mean |     Mean    |     +/-     |     %    |     P    |      Min     |      Max      | 
|:---------------------|--------------:|------------:|------------:|---------:|---------:|-------------:|--------------:|
| startupTime          |         998ms |       975ms |       -22ms |    -2.25 |   BETTER |        965ms |         992ms | 
| vmStartupTime        |         210ms |       204ms |        -6ms |    -2.77 |     SAME |        187ms |         223ms | 
| bgStartupTime        |          48ms |        43ms |        -5ms |   -10.56 |   BETTER |         41ms |          45ms | 
| fgStartupTime        |         649ms |       613ms |       -35ms |    -5.43 |   BETTER |        590ms |         660ms | 
| fgRefreshStartupTime |          92ms |       115ms |        24ms |    25.78 |    WORSE |         86ms |         121ms | 
| fgRestartTime        |           n/a |       NaNms |         n/a |      n/a |      n/a |   Infinityms |   -Infinityms | 
| *totalSize*          |    *29,523kb* |  *59,028kb* |  *29,505kb* |  *99.94* |  *WORSE* |   *57,790kb* |    *59,882kb* | 
| *domSize*            |       *165kb* |     *169kb* |       *3kb* |   *1.90* |  *WORSE* |      *168kb* |       *169kb* | 
| *styleSize*          |       *394kb* |     *396kb* |       *2kb* |   *0.54* |  *WORSE* |      *396kb* |       *399kb* | 
| *jsObjectsSize*      |    *19,757kb* |  *51,387kb* |  *31,631kb* | *160.10* |  *WORSE* |   *51,368kb* |    *51,395kb* | 
| *jsStringsSize*      |       *654kb* |     *500kb* |    *-155kb* | *-23.66* | *BETTER* |      *497kb* |       *501kb* | 
| *jsOtherSize*        |     *8,325kb* |   *6,339kb* |  *-1,986kb* | *-23.85* | *BETTER* |    *5,118kb* |     *7,196kb* | 
| *otherSize*          |       *228kb* |     *237kb* |       *9kb* |   *3.94* |  *WORSE* |      *232kb* |       *239kb* | 
| *USS*                |         *n/a* |     *NaNkb* |       *n/a* |    *n/a* |    *n/a* | *Infinitykb* | *-Infinitykb* | 
| *peakRSS*            |   *366,208kb* | *499,795kb* | *133,587kb* |  *36.48* |  *WORSE* |  *496,976kb* |   *499,952kb* |

### CODE REVIEW

File: int.ts [hunk 1]
```diff
@@ -143,17 +145,15 @@ module J2ME {
     /**
      * Stored in the lower 28 bits.
```
Comment (brendandahl) [inline int.ts]: This comment no longer applies

File: int.ts [hunk 2]
```diff
@@ -143,17 +145,15 @@ module J2ME {
...
      */
-    CalleeMethodInfoOffset      = 2,
-    CalleeMethodInfoMask        = 0x0FFFFFFF,
+    CalleeMethodInfoOffset      = 4,
     /**
      * Stored in the upper 4 bits.
```
Comment (brendandahl) [inline int.ts]: This Comment no longer applies

File: int.ts [hunk 3]
```diff
@@ -99,7 +99,9 @@ module J2ME {
    *             +--------------------------------+
    *             | Caller FP                      |
    *             +--------------------------------+
-   *             | Callee Method Info | Marker    |
+   *             | Frame Type                     |
```
Comment (brendandahl) [inline int.ts]: If we're storing this in a separate slot now we should update the FrameType constants to just use low numbers. (produces faster switch statements when the numbers are dense)

File: int.ts (hunk 2 above)
Comment (mykmelez) [inline int.ts]: D'oh! Fixed in 460e05e, where I also sort and realign the offsets.

File: int.ts (hunk 3 above)
Comment (mykmelez) [inline int.ts]: > If we're storing this in a separate slot now we should update the FrameType constants to just use low numbers. It's easier to debug and produces faster switch statements when the numbers are dense.

Indeed. I had considered combining it with a different value, like the frame pointer, which stays low. But it seemed like a premature optimization. And it's simple to do that later. I've updated the constants to use low numbers in c53c5ef.

**Pull request merged**